File: android/src/main/kotlin/video/api/flutter/player/MethodCallHandler.kt
package video.api.flutter.player

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class MethodCallHandler(
    messenger: BinaryMessenger,
    private val controller: FlutterPlayerController
) :
    MethodChannel.MethodCallHandler {
    private val methodChannel = MethodChannel(messenger, "video.api.player/controller")

    init {
        methodChannel.setMethodCallHandler(this)
    }

    fun dispose() {
        methodChannel.setMethodCallHandler(null)
        controller.disposeAll()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            IS_CREATED -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["isCreated"] = controller.isCreated(it)
                    result.success(reply)
                }
            }
            INITIALIZE -> {
                val autoplay = try {
                    ((call.arguments as Map<*, *>)["autoplay"] as Boolean)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid autoplay", e)
                    return
                }
                val textureId = try {
                    controller.initialize(autoplay)
                } catch (e: Exception) {
                    result.error("failed_action", "Failed to initialize a new player", e)
                    return
                }

                val reply: MutableMap<String, Any> = HashMap()
                reply["textureId"] = textureId
                result.success(reply)
            }
            DISPOSE -> {
                ensureTextureId(call, result) {
                    controller.dispose(it)
                    result.success(null)
                }
            }
            IS_PLAYING -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["isPlaying"] = controller.isPlaying(it)
                    result.success(reply)
                }
            }
            IS_LIVE -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["isLive"] = controller.isLive(it)
                    result.success(reply)
                }
            }
            GET_CURRENT_TIME -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["currentTime"] = controller.getCurrentTime(it)
                    result.success(reply)
                }
            }
            SET_CURRENT_TIME -> {
                val currentTime = try {
                    ((call.arguments as Map<*, *>)["currentTime"] as Int)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid current time", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setCurrentTime(it, currentTime)
                    result.success(null)
                }
            }
            GET_DURATION -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["duration"] = controller.getDuration(it)
                    result.success(reply)
                }
            }
            GET_VIDEO_OPTIONS -> {
                ensureTextureId(call, result) {
                    val videoOptions =
                        controller.getVideoOptions(it)!! // As it is mandatory to set a video option. This should never be null.
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["videoId"] = videoOptions.mediaId
                    reply["type"] = videoOptions.videoType.toFlutterString()
                    result.success(reply)
                }
            }
            SET_VIDEO_OPTIONS -> {
                val videoOptions = try {
                    @Suppress("UNCHECKED_CAST")
                    ((call.arguments as Map<*, *>)["videoOptions"] as Map<String, Any>).videoOptions
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid video options", e)
                    return
                }

                ensureTextureId(call, result) {
                    controller.setVideoOptions(it, videoOptions)
                    result.success(null)
                }
            }
            GET_AUTOPLAY -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["autoplay"] = controller.getAutoplay(it)
                    result.success(reply)
                }
            }
            SET_AUTOPLAY -> {
                val autoplay = try {
                    ((call.arguments as Map<*, *>)["autoplay"] as Boolean)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid autoplay", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setAutoplay(it, autoplay)
                    result.success(null)
                }
            }
            GET_IS_MUTED -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["isMuted"] = controller.getIsMuted(it)
                    result.success(reply)
                }
            }
            SET_IS_MUTED -> {
                val isMuted = try {
                    ((call.arguments as Map<*, *>)["isMuted"] as Boolean)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid isMuted", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setIsMuted(it, isMuted)
                    result.success(null)
                }
            }
            GET_IS_LOOPING -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["isLooping"] = controller.getIsLooping(it)
                    result.success(reply)
                }
            }
            SET_IS_LOOPING -> {
                val isLooping = try {
                    ((call.arguments as Map<*, *>)["isLooping"] as Boolean)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid isLooping", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setIsLooping(it, isLooping)
                    result.success(null)
                }
            }
            GET_VOLUME -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["volume"] = controller.getVolume(it).toDouble()
                    result.success(reply)
                }
            }
            SET_VOLUME -> {
                val volume = try {
                    ((call.arguments as Map<*, *>)["volume"] as Double).toFloat()
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid volume", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setVolume(it, volume)
                    result.success(null)
                }
            }
            GET_VIDEO_SIZE -> {
                ensureTextureId(call, result) {
                    val videoSize = controller.getVideoSize(it)
                    val reply: MutableMap<String, Any> = HashMap()
                    videoSize?.let { size ->
                        reply["width"] = size.width.toDouble()
                        reply["height"] = size.height.toDouble()
                    }
                    result.success(reply)
                }
            }

            SET_PLAYBACK_SPEED -> {
                val speed = try {
                    ((call.arguments as Map<*, *>)["speedRate"] as Double)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid speed", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.setPlaybackSpeed(it, speed)
                    result.success(null)
                }
            }

            GET_PLAYBACK_SPEED -> {
                ensureTextureId(call, result) {
                    val reply: MutableMap<String, Any> = HashMap()
                    reply["speedRate"] = controller.getPlaybackSpeed(it).toDouble()
                    result.success(reply)
                }
            }
            
            PLAY -> {
                ensureTextureId(call, result) {
                    controller.play(it)
                    result.success(null)
                }
            }
            PAUSE -> {
                ensureTextureId(call, result) {
                    controller.pause(it)
                    result.success(null)
                }
            }
            SEEK -> {
                val offset = try {
                    ((call.arguments as Map<*, *>)["offset"] as Int)
                } catch (e: Exception) {
                    result.error("invalid_parameter", "Invalid offset", e)
                    return
                }
                ensureTextureId(call, result) {
                    controller.seek(it, offset)
                    result.success(null)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun ensureTextureId(
        call: MethodCall,
        result: MethodChannel.Result,
        action: (Long) -> Unit
    ) {
        call.argument<Int>(TEXTURE_ID)?.let {
            action(it.toLong())
        } ?: result.error("missing_parameter", "Missing texture Id", null)
    }

    companion object {
        private const val TAG = "ApiVideoPlayerApi"

        private const val TEXTURE_ID = "textureId"

        private const val IS_CREATED = "isCreated"
        private const val INITIALIZE = "initialize"
        private const val DISPOSE = "dispose"

        private const val IS_PLAYING = "isPlaying"
        private const val IS_LIVE = "isLive"
        private const val GET_CURRENT_TIME = "getCurrentTime"
        private const val SET_CURRENT_TIME = "setCurrentTime"
        private const val GET_DURATION = "getDuration"
        private const val GET_VIDEO_OPTIONS = "getVideoOptions"
        private const val SET_VIDEO_OPTIONS = "setVideoOptions"
        private const val GET_AUTOPLAY = "getAutoplay"
        private const val SET_AUTOPLAY = "setAutoplay"
        private const val GET_IS_LOOPING = "getIsLooping"
        private const val SET_IS_LOOPING = "setIsLooping"
        private const val GET_IS_MUTED = "getIsMuted"
        private const val SET_IS_MUTED = "setIsMuted"
        private const val GET_VOLUME = "getVolume"
        private const val SET_VOLUME = "setVolume"
        private const val GET_VIDEO_SIZE = "getVideoSize"
        private const val GET_PLAYBACK_SPEED = "getPlaybackSpeed"
        private const val SET_PLAYBACK_SPEED = "setPlaybackSpeed"

        private const val PLAY = "play"
        private const val PAUSE = "pause"
        private const val SEEK = "seek"
    }
}

File: ios/Classes/MethodCallHandler.swift
import ApiVideoPlayer
import Foundation

class MethodCallHandler {
    private let binaryMessenger: FlutterBinaryMessenger
    private let controller: FlutterPlayerController!
    private let methodChannel: FlutterMethodChannel

    init(binaryMessenger: FlutterBinaryMessenger, controller: FlutterPlayerController) {
        self.binaryMessenger = binaryMessenger
        self.controller = controller
        methodChannel = FlutterMethodChannel(name: "video.api.player/controller", binaryMessenger: binaryMessenger)
        methodChannel
            .setMethodCallHandler { [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) in
                switch call.method {
                case Keys.isCreated:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["isCreated": self?.controller.isCreated(textureId: Int64(textureId))])
                case Keys.initialize:
                    guard let args = call.arguments as? [String: Any],
                          let autoplay = args["autoplay"] as? Bool
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    let textureId = self?.controller.initialize(autoplay: autoplay)
                    result([Keys.textureId: textureId])
                case Keys.dispose:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    self?.controller.dispose(textureId: Int64(textureId))
                    result(nil)
                case Keys.isPlaying:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["isPlaying": self?.controller.isPlaying(textureId: Int64(textureId))])
                case Keys.isLive:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["isLive": self?.controller.isLive(textureId: Int64(textureId))])
                case Keys.getCurrentTime:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["currentTime": self?.controller.getCurrentTime(textureId: Int64(textureId))])
                case Keys.setCurrentTime:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let currentTime = args["currentTime"] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or current time", details: nil))
                        return
                    }
                    self?.controller.setCurrentTime(textureId: Int64(textureId), currentTime: currentTime)
                    result(nil)
                case Keys.getDuration:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["duration": self?.controller.getDuration(textureId: Int64(textureId))])
                case Keys.getVideoOptions:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    let videoOptions = self?.controller.getVideoOptions(textureId: Int64(textureId)) // As it is mandatory to set a video option. This should never be null.
                    result(videoOptions?.toMap())
                case Keys.setVideoOptions:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let videoOptions = args["videoOptions"] as? [String: Any]
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or video options", details: nil))
                        return
                    }
                    self?.controller.setVideoOptions(textureId: Int64(textureId), videoOptions: videoOptions.toVideoOptions())
                    result(nil)
                case Keys.getAutoplay:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["autoplay": self?.controller.getAutoplay(textureId: Int64(textureId))])
                case Keys.setAutoplay:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let autoplay = args["autoplay"] as? Bool
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or autoplay", details: nil))
                        return
                    }
                    self?.controller.setAutoplay(textureId: Int64(textureId), autoplay: autoplay)
                    result(nil)
                case Keys.getIsMuted:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["isMuted": self?.controller.getIsMuted(textureId: Int64(textureId))])
                case Keys.setIsMuted:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let isMuted = args["isMuted"] as? Bool
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or is muted", details: nil))
                        return
                    }
                    self?.controller.setIsMuted(textureId: Int64(textureId), isMuted: isMuted)
                    result(nil)
                case Keys.getIsLooping:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["isLooping": self?.controller.getIsLooping(textureId: Int64(textureId))])
                case Keys.setIsLooping:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let isLooping = args["isLooping"] as? Bool
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or is looping", details: nil))
                        return
                    }
                    self?.controller.setIsLooping(textureId: Int64(textureId), isLooping: isLooping)
                    result(nil)
                case Keys.getVolume:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["volume": self?.controller.getVolume(textureId: Int64(textureId))])
                case Keys.setVolume:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let volume = args["volume"] as? Double
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or volume", details: nil))
                        return
                    }
                    self?.controller.setVolume(textureId: Int64(textureId), volume: Float(volume))
                case Keys.getVideoSize:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    guard let videoSize = self?.controller.getVideoSize(textureId: Int64(textureId)) else {
                        result([String: Any]())
                        return
                    }
                    result(["width": videoSize.width, "height": videoSize.height])
                case Keys.play:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    self?.controller.play(textureId: Int64(textureId))
                    result(nil)
                case Keys.pause:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    self?.controller.pause(textureId: Int64(textureId))
                    result(nil)
                case Keys.seek:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let offset = args["offset"] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or offset", details: nil))
                        return
                    }
                    self?.controller.seek(textureId: Int64(textureId), offset: offset)
                    result(nil)
                case Keys.setPlaybackSpeed:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int,
                          let speed = args["speedRate"] as? Double
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id or speed", details: nil))
                        return
                    }
                    self?.controller.setPlaybackSpeed(textureId: Int64(textureId), speedRate: speed)
                    result(nil)
                case Keys.getPlaybackSpeed:
                    guard let args = call.arguments as? [String: Any],
                          let textureId = args[Keys.textureId] as? Int
                    else {
                        result(FlutterError(code: "invalid_parameter", message: "Failed to get texture id", details: nil))
                        return
                    }
                    result(["speedRate": self?.controller.getPlaybackSpeed(textureId: Int64(textureId))])
                default:
                    result(FlutterMethodNotImplemented)
                }
            }
    }

    func dispose() {
        methodChannel.setMethodCallHandler(nil)
        controller.disposeAll()
    }
}

enum Keys {
    static let textureId = "textureId"

    static let vod = "vod"
    static let live = "live"

    static let isCreated = "isCreated"
    static let initialize = "initialize"
    static let dispose = "dispose"

    static let isPlaying = "isPlaying"
    static let isLive = "isLive"
    static let getCurrentTime = "getCurrentTime"
    static let setCurrentTime = "setCurrentTime"
    static let getDuration = "getDuration"
    static let getVideoOptions = "getVideoOptions"
    static let setVideoOptions = "setVideoOptions"
    static let getAutoplay = "getAutoplay"
    static let setAutoplay = "setAutoplay"
    static let getIsMuted = "getIsMuted"
    static let setIsMuted = "setIsMuted"
    static let getIsLooping = "getIsLooping"
    static let setIsLooping = "setIsLooping"
    static let getVolume = "getVolume"
    static let setVolume = "setVolume"
    static let getVideoSize = "getVideoSize"
    static let getPlaybackSpeed = "getPlaybackSpeed"
    static let setPlaybackSpeed = "setPlaybackSpeed"

    static let play = "play"
    static let pause = "pause"
    static let seek = "seek"
}

extension String {
    func toVideoType() -> VideoType {
        switch self {
        case Keys.vod:
            return VideoType.vod
        case Keys.live:
            return VideoType.live
        default:
            return VideoType.vod
        }
    }
}

extension VideoType {
    func toString() -> String {
        switch self {
        case VideoType.vod:
            return Keys.vod
        case VideoType.live:
            return Keys.live
        }
    }
}

extension Dictionary where Key == String {
    func toVideoOptions() -> VideoOptions {
        return VideoOptions(videoId: self["videoId"] as! String, videoType: (self["type"] as! String).toVideoType(), token: self["token"] as? String)
    }
}

extension VideoOptions {
    func toMap() -> [String: Any?] {
        return [
            "videoId": videoId,
            "type": videoType.toString(),
            "token": token ?? nil,
        ]
    }
}
